{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Multiple Sequence Alignment using Multi-objective Genetic Algorithms\n",
    "\n",
    "### Authors:\n",
    "1. Urvish Vasani - uvasani\n",
    "2. Kenil Shah - kshah9\n",
    "3. Suraj Patel - sjpate22\n",
    "4. Priyanka Hebsur - phebsur\n",
    "\n",
    "### Install external dependencies\n",
    "Execute this block below to install required libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m\u001b[1m   Updating\u001b[22m\u001b[39m registry at `C:\\Users\\ur207\\.julia\\registries\\General`\n",
      "┌ Warning: could not download https://pkg.julialang.org/registries\n",
      "└ @ Pkg.Types D:\\buildbot\\worker\\package_win64\\build\\usr\\share\\julia\\stdlib\\v1.5\\Pkg\\src\\Types.jl:951\n",
      "\u001b[32m\u001b[1m  Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `C:\\Users\\ur207\\.julia\\environments\\v1.5\\Project.toml`\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `C:\\Users\\ur207\\.julia\\environments\\v1.5\\Manifest.toml`\n",
      "\u001b[32m\u001b[1m  Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `C:\\Users\\ur207\\.julia\\environments\\v1.5\\Project.toml`\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `C:\\Users\\ur207\\.julia\\environments\\v1.5\\Manifest.toml`\n",
      "\u001b[32m\u001b[1m  Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `C:\\Users\\ur207\\.julia\\environments\\v1.5\\Project.toml`\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `C:\\Users\\ur207\\.julia\\environments\\v1.5\\Manifest.toml`\n",
      "\u001b[32m\u001b[1m  Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `C:\\Users\\ur207\\.julia\\environments\\v1.5\\Project.toml`\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `C:\\Users\\ur207\\.julia\\environments\\v1.5\\Manifest.toml`\n"
     ]
    }
   ],
   "source": [
    "import Pkg\n",
    "Pkg.add(\"FASTX\")\n",
    "Pkg.add(\"Distributions\")\n",
    "Pkg.add(\"BioSequences\")\n",
    "Pkg.add(\"Plots\")\n",
    "\n",
    "using FASTX\n",
    "using Distributions\n",
    "using BioSequences\n",
    "using Random\n",
    "using Plots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Chromosome Codification\n",
    "\n",
    "<hr>\n",
    "\n",
    "## Function: codify_chromosome\n",
    "This function takes an individual which is list of strings(sequence) and converts into a encoded form. For efficiently implementing the crossover operator and mutation operator, this approach represents alignments as a two-dimensional matrix where each row stands for each alignment and includes the following two conditions: \n",
    "* amino acids are coded with their position in the sequence where they belong\n",
    "* gaps are coded with the ‘negative’ position of the last amino acid in the sequence\n",
    "\n",
    "### Input :\n",
    "* msa (List of sequences)\n",
    "\n",
    "### Output :\n",
    "* chromosom (Encoded MSA)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "codify_chromosome (generic function with 1 method)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function codify_chromosome(msa)\n",
    "    chromosome = [[0 for j in 1:length(msa[1])] for i in 1:length(msa)] \n",
    "    for i in 1:length(msa)\n",
    "        count = 0\n",
    "        for j in 1:length(msa[i])\n",
    "            if msa[i][j] == '-'\n",
    "                chromosome[i][j] = count*(-1) \n",
    "            else\n",
    "                count = count + 1\n",
    "                chromosome[i][j] = count\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    return chromosome\n",
    "end\n",
    "#codify_chromosome([\"ABCDEFGH--\",\"AB----CDEF\",\"ABCD----EF\"])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Helper Functions\n",
    "<hr>\n",
    "\n",
    "## function: remove_columns_with_all_gaps\n",
    "This function is used to remove the columns with all gaps and no nucleotides. This can happen as a side effect of the crossover and mutation operations.\n",
    "\n",
    "### Input:\n",
    "* sequences (Alignments having columns with gaps)\n",
    "* codifications (Alignment codification)\n",
    "\n",
    "### Output:\n",
    "* sequences (Alignments without any column having all gaps)\n",
    "* codifications (Modified codification)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "remove_columns_with_all_gaps (generic function with 1 method)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function remove_columns_with_all_gaps(sequences, codifications)\n",
    "    columns_with_all_gaps = []\n",
    "    total_seq = length(sequences)\n",
    "    \n",
    "    for (index,nucleotide) in enumerate(sequences[1])\n",
    "        if nucleotide == '-'\n",
    "            column_with_all_gaps = true\n",
    "            \n",
    "            for i in 2:total_seq\n",
    "                if sequences[i][index] != '-'\n",
    "                    column_with_all_gaps = false\n",
    "                    break\n",
    "                end\n",
    "            end\n",
    "            \n",
    "            if column_with_all_gaps\n",
    "                append!(columns_with_all_gaps, index)\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    for i in 1:total_seq\n",
    "        sequences[i] = join([c for (i,c) in enumerate(sequences[i]) if ~(i in columns_with_all_gaps)])\n",
    "        codifications[i] = [c for (i,c) in enumerate(codifications[i]) if ~(i in columns_with_all_gaps)]\n",
    "    end\n",
    "                    \n",
    "    return sequences, codifications\n",
    "end\n",
    "                                \n",
    "#remove_columns_with_all_gaps([\"ab---fg\",\"ab---fg\",\"ab-c-fg\",\"ab---fg\"], [[1,2,-2,-2,-2,3,4],[1,2,-2,-2,-2,3,4],[1,2,-2,3,-3,4,5],[1,2,-2,-2,-2,3,4]])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Crossover Operation\n",
    "\n",
    "---\n",
    "\n",
    "## function: get_cut_point\n",
    "This function would take in the elements found after the cut and the encoded version of the sequence and it would return us the cut_point index for the encoded sequence\n",
    "\n",
    "### Input:\n",
    "* cut_elements (Elements which are found at the cut point for string 1)\n",
    "* chromosome_coded (Encoded sequence of string 2)\n",
    "\n",
    "### Output:\n",
    "* cut_point: (Cut point index for string 2)\n",
    "\n",
    "***\n",
    "\n",
    "## function: crossover\n",
    "This function would take two parent candidate strings and there encoded version in order to generate two children by using the strategy of crossover. In crossover the cut elements of both the candidates are matched and then children are generated by merging left/right and right/left part of the both the parents.\n",
    "\n",
    "### Input:\n",
    "* chromosome_string1 (Parent 1 which will have a list of sequence strings)\n",
    "* chromosome_string2 (Parent 2 which will have a list of sequence strings)\n",
    "* chromosome_coded1  (Encoded version of first parent)\n",
    "* chromosome_coded2  (Encoded version of second parent)\n",
    "\n",
    "### Output:\n",
    "* crossover_string (Two child strings generated by two different combination of pairs)\n",
    "* crossover_coded  (Two encoded version of the children)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "crossover (generic function with 1 method)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function get_cut_point(cut_elements,chromosome_coded)\n",
    "    cut_point = 0\n",
    "    for i in 1:length(chromosome_coded)\n",
    "        if abs(chromosome_coded[i])==abs(cut_elements[1])\n",
    "            cut_point = i\n",
    "            break\n",
    "        end\n",
    "    end\n",
    "    return cut_point\n",
    "end\n",
    "\n",
    "function crossover(chromosome_string1, chromosome_coded1,chromosome_string2, chromosome_coded2)\n",
    "    minimum = min(length(chromosome_string1[1]), length(chromosome_string2[1]))\n",
    "    cut_point = rand(2:minimum-1)\n",
    "\n",
    "    diff1,diff2 = Inf,-Inf\n",
    "    \n",
    "    cut_points_string2 = []\n",
    "    \n",
    "    for i in 1:length(chromosome_string1)\n",
    "        cut_elements =  [chromosome_coded1[i][cut_point],chromosome_coded1[i][cut_point+1]]\n",
    "        cut_point2 = get_cut_point(cut_elements,chromosome_coded2[i])\n",
    "        diff1 = min(diff1,cut_point2)\n",
    "        diff2 = max(diff2,cut_point2)\n",
    "        push!(cut_points_string2,cut_point2)\n",
    "    end\n",
    "    \n",
    "    crossover_string = [[],[]]\n",
    "    crossover_coded = [[],[]]\n",
    "    \n",
    "    for i in 1:length(chromosome_string1)\n",
    "        cut_point2 = cut_points_string2[i]\n",
    "        s1 = chromosome_string1[i][1:cut_point]\n",
    "        s2 = chromosome_string2[i][cut_point2+1:length(chromosome_string2[i])]\n",
    "        \n",
    "        gap_count1 = Int(cut_point2 - diff1)\n",
    "        gaps1 = \"-\"^gap_count1\n",
    "        \n",
    "        main_string1 = string(s1,gaps1,s2)        \n",
    "        main_coded1 = codify_chromosome([main_string1])[1]\n",
    "        \n",
    "        s1 = chromosome_string2[i][1:cut_point2]\n",
    "        s2 = chromosome_string1[i][cut_point+1:length(chromosome_string1[i])]\n",
    "\n",
    "        gap_count2 = Int(diff2 - cut_point2)\n",
    "        gaps2 = \"-\"^gap_count2\n",
    "\n",
    "        main_string2 = string(s1,gaps2,s2)        \n",
    "        main_coded2 = codify_chromosome([main_string2])[1]\n",
    "        \n",
    "        push!(crossover_string[1], main_string1)\n",
    "        push!(crossover_coded[1], main_coded1)\n",
    "        \n",
    "        push!(crossover_string[2], main_string2)\n",
    "        push!(crossover_coded[2], main_coded2)\n",
    "    end\n",
    "    crossover_string[1],crossover_coded[1] = remove_columns_with_all_gaps(crossover_string[1],crossover_coded[1])\n",
    "    crossover_string[2],crossover_coded[2] = remove_columns_with_all_gaps(crossover_string[2],crossover_coded[2])\n",
    "\n",
    "    return crossover_string,crossover_coded\n",
    "    \n",
    "end\n",
    "\n",
    "#crossover([\"ACGTT--AC-G-GG-\", \"A-GGCTTTAGG-CG-\", \"AG-G-CT-ATGCAGG\"],[[1, 2, 3, 4, 5, -5, -5, 6, 7, -7, 8, -8, 9, 10, -10], [1, -1, 2, 3, 4, 5, 6, 7, 8, 9, 10, -10, 11, 12, -12], [1, 2, -2, 3, -3, 4, 5, -5, 6, 7, 8, 9, 10, 11, 12]],[\"AC--G--TTA-CGGG\", \"AGGC-T-T-TAGGCG\", \"A-GGCTATGC-A-GG\"],[[1, 2, -2, -2, 3, -3, -3, 4, 5, 6, -6, 7, 8, 9, 10], [1, 2, 3, 4, -4, 5, -5, 6, -6, 7, 8, 9, 10, 11, 12], [1, -1, 2, 3, 4, 5, 6, 7, 8, 9, -9, 10, -10, 11, 12]])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Mutation Operation\n",
    "\n",
    "<hr>\n",
    "\n",
    "## function: mutation\n",
    "\n",
    "This function takes an individual alignment and its codification as input and generated the offspring. For our algorithm, mutation operator moves the group of gaps from one position to other in order to generate the offspring.\n",
    "\n",
    "### input:\n",
    "* parent_sequence (parent consists of the list of strings)\n",
    "* parent_alignment (codification of the parent_sequence)\n",
    "\n",
    "### output:\n",
    "* child_sequence (Generated Offspring as mentioned above)\n",
    "* child_codification (codification of the generated child)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mutation (generic function with 1 method)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function mutation(parent_sequence, parent_codification)\n",
    "    child_sequence = []\n",
    "    child_codification = []\n",
    "    \n",
    "    for i in 1:length(parent_codification)\n",
    "        curr_codification = parent_codification[i]\n",
    "        curr_sequence = parent_sequence[i]\n",
    "        \n",
    "        gaps = []\n",
    "        gap_start = -1\n",
    "        for j in 1:length(curr_sequence)\n",
    "            if curr_sequence[j] == '-'\n",
    "                if gap_start == -1\n",
    "                    gap_start = j\n",
    "                end\n",
    "            else\n",
    "                if gap_start != -1\n",
    "                    push!(gaps,[gap_start,j-1])\n",
    "                    gap_start = -1\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "        if gap_start != -1\n",
    "            push!(gaps,[gap_start,length(curr_sequence)])\n",
    "        end\n",
    "        \n",
    "        if length(gaps) == 0\n",
    "            mutated_seq = join([c for (i,c) in enumerate(curr_sequence)])\n",
    "            mutated_codification = [c for (i,c) in enumerate(curr_codification)]\n",
    "            \n",
    "            push!(child_sequence, mutated_seq)\n",
    "            push!(child_codification, mutated_codification)\n",
    "            \n",
    "            continue\n",
    "        end\n",
    "        \n",
    "        gap_group = gaps[rand(1:length(gaps))]\n",
    "        \n",
    "        mutated_seq = join([c for (i,c) in enumerate(curr_sequence) if ~(i>=gap_group[1] && i<=gap_group[2])])\n",
    "        mutated_codification = [c for (i,c) in enumerate(curr_codification) if ~(i>=gap_group[1] && i<=gap_group[2])]\n",
    "                        \n",
    "        target_position = rand(1:length(mutated_seq)+1)\n",
    "        \n",
    "        gap_sequence = join(['-' for i in gap_group[1]:gap_group[2]])\n",
    "        \n",
    "        if target_position == length(mutated_seq)+1\n",
    "            mutated_seq = mutated_seq * gap_sequence\n",
    "            gap_code = -1*abs(mutated_codification[length(mutated_codification)])\n",
    "            mutated_codification = cat(mutated_codification,[gap_code for i in gap_group[1]:gap_group[2]], dims=1)\n",
    "        elseif target_position == 1\n",
    "            mutated_seq = gap_sequence * mutated_seq\n",
    "            gap_code = 0\n",
    "            mutated_codification = cat([gap_code for i in gap_group[1]:gap_group[2]], mutated_codification, dims=1)\n",
    "        else\n",
    "            mutated_seq = mutated_seq[1:target_position-1] * gap_sequence * mutated_seq[target_position:length(mutated_seq)]\n",
    "            gap_code = -1*abs(mutated_codification[target_position-1])\n",
    "            mutated_codification = cat(mutated_codification[1:target_position-1],[gap_code for i in gap_group[1]:gap_group[2]],mutated_codification[target_position:length(mutated_codification)],dims=1)          \n",
    "        end\n",
    "                                    \n",
    "        push!(child_sequence, mutated_seq)\n",
    "        push!(child_codification, mutated_codification)\n",
    "    end\n",
    "    \n",
    "    child_sequence,child_codification = remove_columns_with_all_gaps(child_sequence, child_codification)\n",
    "end\n",
    "\n",
    "#mutation([\"abcd---cd--ef-cde-\", \"abcd---cd--ef-cdef\"], [[1,2,3,4,-4,-4,-4,5,6,-6,-6,7,8,-8,9,10,11,-11],[1,2,3,4,-4,-4,-4,5,6,-6,-6,7,8,-8,9,10,11,12]])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Initial Population Generation\n",
    "***\n",
    "\n",
    "## Function: add_initial_gaps\n",
    "This function simply adds gaps to a sequence\n",
    "\n",
    "### Input :\n",
    "* Gap Count (No. of gaps to be added)\n",
    "* Single Sequence\n",
    "\n",
    "### Output :\n",
    "* Sequence with gaps in it\n",
    "\n",
    "---\n",
    "## Function: generate_single_chromosome\n",
    "This function would take in the multiple sequences and will try to add some number of gaps to every sequence by following\n",
    "the below mentioned instructions:\n",
    "* Find the sequence with maximum length in the multiple sequennces\n",
    "* Add gaps ranging from randomly 20-40% in this specific sequence\n",
    "* Since we have to make every sequence of same length, we would compute the number of gaps needed to add in every sequence\n",
    "to make it similar to the sequence with maximum length.\n",
    "\n",
    "### Input :\n",
    "* Chromosome (Multiple Sequence)\n",
    "\n",
    "### Output :\n",
    "* Chromosome (Multiple Sequences with gaps added to them)\n",
    "\n",
    "***\n",
    "\n",
    "## Function: generate_initial_population\n",
    "This function generates a set of chromosomes after adding gaps to given initial sequence. After adding 20% of chromosomes with gaps in it, we would crossover those sequences in order to generate other 80% of chromosomes. This set of chromosomes would be our initial population.\n",
    "\n",
    "### Input :\n",
    "* Chromosome (Multiple Sequence)\n",
    "* population_count (Number of chromosomes for initial population)\n",
    "\n",
    "### Output :\n",
    "* population_string (Will contain population_count number of candidates)\n",
    "* population_coded  (Will contain population_count number of encoded candidates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "generate_initial_population (generic function with 1 method)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function add_initial_gaps(gap_count,sequence)\n",
    "    gap_count = floor(gap_count)\n",
    "    while gap_count!=0\n",
    "        blank_count = Int(rand(1:gap_count))\n",
    "        position  = rand(1:length(sequence))\n",
    "        sequence = string(sequence[1:position],\"-\"^blank_count,sequence[position+1:length(sequence)])\n",
    "        gap_count=gap_count-blank_count\n",
    "    end\n",
    "    return sequence\n",
    "end\n",
    "\n",
    "function generate_single_chromosome(msa)\n",
    "    \n",
    "    new_msa = []\n",
    "    for i in 1:length(msa)\n",
    "        push!(new_msa,\"\")\n",
    "    end\n",
    "    \n",
    "    max_sequence_length = length(msa[1])\n",
    "    max_sequence_index = 1\n",
    "    for (i,sequence) in enumerate(msa)\n",
    "        if length(sequence)>max_sequence_length\n",
    "            max_sequence_length = length(sequence)\n",
    "            max_sequence_index = i\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    # Update Max Length Sequence\n",
    "    \n",
    "    gap_count = rand(Uniform(0.2,0.4)) * max_sequence_length\n",
    "    new_msa[max_sequence_index] = add_initial_gaps(gap_count,msa[max_sequence_index])\n",
    "    \n",
    "    # Update Rest of the Sequences\n",
    "    \n",
    "    for i in 1:length(msa)\n",
    "        if i==max_sequence_index\n",
    "            continue\n",
    "        end\n",
    "        gap_counts = length(new_msa[max_sequence_index]) - length(msa[i])\n",
    "        new_msa[i] = add_initial_gaps(gap_counts,msa[i])\n",
    "    end\n",
    "    return new_msa\n",
    "end\n",
    "\n",
    "function generate_initial_population(msa,population_count)        \n",
    "    without_crossover = 0.2*population_count\n",
    "    population_string = []\n",
    "    population_coded = []\n",
    "    for i in 1:without_crossover\n",
    "        chromosome = generate_single_chromosome(msa)\n",
    "        coded_chromosome = codify_chromosome(chromosome)\n",
    "        chromosome,coded_chromosome = remove_columns_with_all_gaps(chromosome,coded_chromosome)\n",
    "        push!(population_coded,coded_chromosome)\n",
    "        push!(population_string,chromosome)\n",
    "    end\n",
    "        \n",
    "    with_crossover = (0.8*population_count)/2\n",
    "    for i in 1:with_crossover\n",
    "        random1 = Int(rand(1:without_crossover))\n",
    "        random2 = Int(rand(1:without_crossover))\n",
    "        crossover_string, crossover_coded = crossover(population_string[random1],population_coded[random1],population_string[random2],population_coded[random2])\n",
    "        population_string = vcat(population_string,crossover_string)\n",
    "        population_coded = vcat(population_coded,crossover_coded)\n",
    "    end\n",
    "    return population_string, population_coded\n",
    "end    \n",
    "\n",
    "#generate_initial_population([\"ACGTTACCCGGAAATTTTTTACGGG\",\"ACGTTACCTGGGAATATTGTACAGG\",\"GTGTAAGGTGGGAATATTTFAGAGG\"],100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Fitness functions\n",
    "***\n",
    "\n",
    "## Function : compute_tc_score\n",
    "Given an input sequence, this function calculates the percentage of columns which are completely aligned with same amino acids. \n",
    "\n",
    "### Input:\n",
    "* sequences (individual alignment)\n",
    "\n",
    "### Output:\n",
    "* score (Percentage of totally consered columns)\n",
    "\n",
    "***\n",
    "\n",
    "## Function : calc_sum_pair\n",
    "Given an input sequence, this function returns the sum of pairwise alignment score. Bonus for match is 1, mismatch penalty is zero and Gap penalty is zero.\n",
    "\n",
    "### Input:\n",
    "* sequences (individual alignment)\n",
    "\n",
    "### Output:\n",
    "* score (Percentage of totally consered columns)\n",
    "\n",
    "***\n",
    "\n",
    "## Function : calc_nogap_percentage\n",
    "In order to penalize overuse of gaps in the produced alignment, this function calculates the percentage of amino acids in the given alignment as compared to the number of gaps.\n",
    "\n",
    "### Input:\n",
    "* sequences (individual alignment)\n",
    "\n",
    "### Output:\n",
    "* no_gaps_score (Percentage of totally consered columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "calculate_fitness_score (generic function with 1 method)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function compute_tc_score(sequences)\n",
    "    no_of_aligned_columns = 0\n",
    "    total_columns = length(sequences[1])\n",
    "    is_aligned = true\n",
    "    for i in 1:total_columns\n",
    "        is_aligned = true\n",
    "        first_residue= sequences[1][i]\n",
    "        if  first_residue != '-'\n",
    "            for j in 1:length(sequences)\n",
    "                if first_residue != sequences[j][i]\n",
    "                    is_aligned = false\n",
    "                    break\n",
    "                end\n",
    "            end\n",
    "            else is_aligned = false\n",
    "        end\n",
    "        if is_aligned \n",
    "            no_of_aligned_columns += 1\n",
    "        end\n",
    "    end\n",
    "    score = (100.0 * no_of_aligned_columns) / total_columns\n",
    "    return score \n",
    "end\n",
    "\n",
    "# +1 for match and 0 for no match\n",
    "\n",
    "function calc_sum_pair(sequences) \n",
    "    t = length(sequences)\n",
    "    k = length(sequences[1])\n",
    "    score = 0\n",
    "    for i=1:t\n",
    "        A = sequences[i]\n",
    "        for j=i+1:t\n",
    "            B = sequences[j]\n",
    "            for idx = 1:k\n",
    "                if A[idx] == B[idx] && A[idx] != '-'\n",
    "                    score += 1\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    return score\n",
    "end\n",
    "\n",
    "function calc_nogap_percentage(sequences)\n",
    "    t = length(sequences)\n",
    "    k = length(sequences[1])\n",
    "    no_gaps = 0\n",
    "    for i in 1:t\n",
    "        for j in 1:k\n",
    "            if sequences[i][j]!='-'\n",
    "                no_gaps+=1\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    total = k*t\n",
    "    no_gaps_score = (no_gaps/total)*100\n",
    "    return no_gaps_score\n",
    "end\n",
    "    \n",
    "# +1 match. -1 mismatch and -2 for gaps\n",
    "function calc_sum_pair2(sequences) \n",
    "    t = length(sequences)\n",
    "    k = length(sequences[1])\n",
    "    score = 0\n",
    "    for i=1:t\n",
    "        A = sequences[i]\n",
    "        for j=i+1:t\n",
    "            B = sequences[j]\n",
    "            for idx = 1:k\n",
    "                if A[idx]=='-' || B[idx]=='-'\n",
    "                    score -=2\n",
    "                elseif A[idx] == B[idx]\n",
    "                    score += 1\n",
    "                else\n",
    "                    score -= 1\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    return score\n",
    "end\n",
    "\n",
    "function calculate_fitness_score(align)\n",
    "    return [compute_tc_score(align), calc_sum_pair(align), calc_nogap_percentage(align), calc_sum_pair2(align)]\n",
    "end\n",
    "#Test \n",
    "#sequences = [\"AGGCTTT-A-C\", \"CCCAGTG-AT-\",\"GGCFATT-AT-\"]\n",
    "#print(calculate_fitness_score(sequences))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. NSGA-II algorithm for multi-objectve optimization\n",
    "\n",
    "*** \n",
    "Matlab implementation of the paper used inbuilt functioalities to execute NSGA-II. Since there is no such reliable support present in julia as of now, we implemented our own NSGA with the help of reference:  https://github.com/gsoleilhac/NSGAII.jl . Authors of this repository basically implemented the NSGA-II algorithm to solve the knapsack problem.\n",
    "\n",
    "***\n",
    "\n",
    "## Structure: Individual\n",
    "This structure is used to store the information about each individual in the current population. In particular, following information is stored:\n",
    "* align : list of string representing the alignment\n",
    "* codification : codification of the alignment present in the align\n",
    "* y : list of fitness score values\n",
    "* rank : rank of the pareto front assigned to this individual\n",
    "* crowding : crowding distance for this individual\n",
    "* dom_count : Number of individuls that dominated this particular individual\n",
    "* dom_list : list of individuals dominated by this individual\n",
    "\n",
    "***\n",
    "\n",
    "## Function: dominates\n",
    "This function returns whether first individual dominate the other in terms of fitness scores.\n",
    "\n",
    "### input\n",
    "* a : First individual\n",
    "* b : Seconf individual\n",
    "\n",
    "### output\n",
    "* result : true if first individual dominates the other and false otherwise\n",
    "\n",
    "***\n",
    "\n",
    "## Function: isless\n",
    "This function returns whether the rank of the first individual is lower than the other. In case where rank of both the individuals are the same, crowding distace is used.\n",
    "\n",
    "### input\n",
    "* a : First individual\n",
    "* b : Seconf individual\n",
    "\n",
    "### output\n",
    "* result : true if rank of the first individual is lower than the second"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "isless (generic function with 1 method)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Basic structure to store the individuals and helper functions\n",
    "\n",
    "mutable struct Individual{Align, Codification, Fitness}\n",
    "    align::Align\n",
    "    codification::Codification\n",
    "    y::Fitness\n",
    "    rank::UInt16\n",
    "    crowding::Float64\n",
    "    dom_count::UInt16\n",
    "    dom_list::Vector{UInt16}\n",
    "    \n",
    "    Individual(align::Align, codification::Codification ,y::Fitness) where {Align,Codification,Fitness} = new{Align,Codification,Fitness}(align, codification, y,zero(UInt16),0.,zero(UInt16),UInt16[]) \n",
    "end\n",
    "\n",
    "function dominates(a::Individual, b::Individual)\n",
    "    result = false\n",
    "    for i in eachindex(a.y)\n",
    "        a.y[i] < b.y[i] && return false\n",
    "        a.y[i] > b.y[i] && (result = true)\n",
    "    end\n",
    "    return result\n",
    "end\n",
    "\n",
    "function isless(a::Individual, b::Individual)\n",
    "    return a.rank < b.rank || a.rank == b.rank && a.crowding >= b.crowding\n",
    "end\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function: calculate_Pareto_Fronts!\n",
    "This function graphs the pareto front on the individuals present inside population and assigns rank to each individual as well according to the front in which they belong.\n",
    "\n",
    "### Input\n",
    "* population : List of Individuals present in the current population\n",
    "\n",
    "### Output\n",
    "* This function does not return anything but assigns the ranks to the individuals in-place."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "calculate_Pareto_Fronts! (generic function with 1 method)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function calculate_Pareto_Fronts!(population)\n",
    "    n = length(population)\n",
    "    for p in population\n",
    "        empty!(p.dom_list)\n",
    "        p.dom_count = 0\n",
    "        p.rank = 0\n",
    "    end\n",
    "    \n",
    "    ranked_individuals = 0\n",
    "    \n",
    "    current_rank_index = []\n",
    "    for i in 1:n\n",
    "        for j in i+1:n\n",
    "            if dominates(population[i], population[j])\n",
    "                push!(population[i].dom_list, j)\n",
    "                population[j].dom_count += 1\n",
    "            elseif dominates(population[j], population[i])\n",
    "                push!(population[j].dom_list, i)\n",
    "                population[i].dom_count += 1\n",
    "            end\n",
    "        end\n",
    "        if population[i].dom_count == 0\n",
    "            population[i].rank = 1\n",
    "            ranked_individuals += 1\n",
    "            append!(current_rank_index, i)\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    current_rank = UInt16(2)\n",
    "    while ranked_individuals != n\n",
    "        indiv_to_delete = deepcopy(current_rank_index)\n",
    "        empty!(current_rank_index)\n",
    "        \n",
    "        for i in indiv_to_delete\n",
    "            p = population[i]\n",
    "\n",
    "            for q in p.dom_list\n",
    "                population[q].dom_count -= one(UInt16)\n",
    "                if population[q].dom_count == zero(UInt16)\n",
    "                    population[q].rank = current_rank\n",
    "                    ranked_individuals += 1\n",
    "                    append!(current_rank_index, q)\n",
    "                end\n",
    "            end\n",
    "\n",
    "        end\n",
    "        current_rank += one(UInt16)        \n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function: tournament_selection\n",
    "This function randomly samples two individuals from the whole population and returns the one with lower rank and higher crowding distance.\n",
    "\n",
    "### input\n",
    "* P : population of individuals\n",
    "\n",
    "### output\n",
    "* Individual : returns the individual sampled from the above mentioned approach."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tournament_selection (generic function with 1 method)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function tournament_selection(P)\n",
    "    a,b = rand(1:length(P)÷2), rand(1:length(P)÷2)\n",
    "    if isless(P[a], P[b])\n",
    "        return P[a]\n",
    "    else\n",
    "        return P[b]\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function: crowding_distance_sorting!\n",
    "This function computed the crowding distance for the population passed as the input and assigns distance value to each individuals present inside the population.\n",
    "\n",
    "### Input\n",
    "* population : List of Individuals present in the current population\n",
    "\n",
    "### Output\n",
    "* This function does not return anything but assigns the crowding distance to the individuals in-place."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "crowding_distance_sorting! (generic function with 1 method)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function crowding_distance_sorting!(population)\n",
    "    for p in population\n",
    "        p.crowding = 0.\n",
    "    end\n",
    "    for j = 1:length(first(population).y)\n",
    "        sort!(population, by = x -> x.y[j])\n",
    "        population[1].crowding = population[end].crowding = Inf\n",
    "        if population[1].y[j] != population[end].y[j]\n",
    "            for i in 2:length(population)-1\n",
    "                population[i].crowding += (population[i+1].y[j] - population[i-1].y[j]) / (population[end].y[j] - population[1].y[j])\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function: NSGA\n",
    "This function executes the discussed NSGA approach on the provided input sequence.\n",
    "\n",
    "### Input\n",
    "* input_seq : Input alignments\n",
    "* population_size : size of the population\n",
    "* num_of_gen : number of generations\n",
    "* pmut : probability of mutation\n",
    "\n",
    "### steps\n",
    "* From the provided input sequence, initial population of population_size is generated\n",
    "* For this initial population, pareto fronts are generated and assigned ranks as well.\n",
    "* Now, for each generation, population is doubled using crossover and mutation operations. Then, pareto front is generated and best individuals are included in the next population.\n",
    "* At the end, best individual from the first pareto front is returned.\n",
    "\n",
    "### output\n",
    "* best_sol : Individual with best alignment score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NSGA (generic function with 1 method)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function NSGA(input_seq, population_size, num_of_gen, pmut)\n",
    "    alignments, codes = generate_initial_population(input_seq, population_size)\n",
    "    \n",
    "    P = Vector{Individual}(undef, 2*population_size)    # Whole population \n",
    "    \n",
    "    for i in 1:population_size                                # Create Initial population\n",
    "        P[i] = Individual(alignments[i],codes[i], calculate_fitness_score(alignments[i])) \n",
    "        P[population_size+i] = deepcopy(P[i])\n",
    "    end\n",
    "    \n",
    "    calculate_Pareto_Fronts!(view(P, 1:population_size))\n",
    "    \n",
    "    for gen in 1:num_of_gen\n",
    "        for i = 1:2:population_size\n",
    "            \n",
    "            pa = tournament_selection(P)\n",
    "            pb = tournament_selection(P)\n",
    "            \n",
    "            childs_align, childs_codification = crossover(pa.align, pa.codification, pb.align, pb.codification)\n",
    "            \n",
    "            if rand() < pmut\n",
    "                childs_align[1],childs_codification[1] = mutation(childs_align[1],childs_codification[1])\n",
    "            end\n",
    "                \n",
    "            if rand() < pmut\n",
    "                childs_align[2], childs_codification[2] = mutation(childs_align[2],childs_codification[2])\n",
    "            end\n",
    "            \n",
    "            P[population_size+i] = Individual(childs_align[1],childs_codification[1], calculate_fitness_score(childs_align[1])) \n",
    "            P[population_size+i+1] = Individual(childs_align[2],childs_codification[2], calculate_fitness_score(childs_align[2])) \n",
    "            \n",
    "        end\n",
    "        calculate_Pareto_Fronts!(P)\n",
    "        \n",
    "        sort!(P, by = x -> x.rank, alg = Base.Sort.QuickSort)\n",
    "        \n",
    "        let f::Int = 1\n",
    "            ind = 0\n",
    "            indnext = findlast(x -> x.rank == f, P)\n",
    "            while 0 < indnext <= population_size\n",
    "                ind = indnext\n",
    "                f += 1\n",
    "                indnext = findlast(x -> x.rank == f, P)\n",
    "            end\n",
    "            indnext == 0 && (indnext = length(P))\n",
    "            crowding_distance_sorting!(view(P, ind+1:indnext))\n",
    "            sort!(view(P, (ind + 1):indnext), by = x -> x.crowding, rev = true, alg = PartialQuickSort(population_size - ind))\n",
    "            \n",
    "        end\n",
    "        \n",
    "    end\n",
    "    optimal_sols = filter(x -> x.rank == 1, view(P, 1:population_size))\n",
    "    \n",
    "    best_sol = optimal_sols[1]\n",
    "    for i in 2:length(optimal_sols)\n",
    "        if optimal_sols[i].y[2] > best_sol.y[2]\n",
    "             best_sol = optimal_sols[i]\n",
    "        end \n",
    "    end\n",
    "    \n",
    "    return best_sol\n",
    "    \n",
    "end\n",
    "\n",
    "#input_seq = [\"AATGGTCATAGCGAGATGAAGCCACGTGATGGATAATATTGTGCAAACGACCTTATTAGCTATTGACCGTCGATGTCCAACGAGACAATT\", \"GAATCTGTATTCTTCAAGCTTCAACTCCATGCACTACGAACGGTAGTGGTTCACATTGACCGTG\", \"TTGGGCGCATTGACCGTCCTTCCTAGCGTATCATCAAACTTGTGATTCTCTATCTAGAGCAAAATGCGGTGTCCGCTATATGGAGATCTATTTCAAAATA\", \"GTTACATATCAGAATCATTAGAAACGCTCTTAATGGGGTTAAGCAGAGACTTAGTAAGGATTAACTCCCAAGATGATTGACCGTGCTC\"]\n",
    "#NSGA(input_seq,100,300,0.8)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 8. reading input from the file\n",
    "\n",
    "***\n",
    "\n",
    "## Function: read_input\n",
    "\n",
    "this function reads file with input sequences in the FASTA format and returns the identifiers and sequences.\n",
    "\n",
    "### Input\n",
    "* filename : name of the input file\n",
    "\n",
    "### Output\n",
    "* identifiers : list of identifiers \n",
    "* sequences : list of input sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "read_input (generic function with 1 method)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read sequences from the input file\n",
    "# Reference: https://biojulia.net/FASTX.jl/dev/manual/fasta/\n",
    "\n",
    "function read_input(filename)\n",
    "    reader = open(FASTA.Reader, filename)\n",
    "    \n",
    "    identifiers = []\n",
    "    sequences = []\n",
    "    \n",
    "    for record in reader\n",
    "        push!(identifiers, string(FASTA.identifier(record)))\n",
    "        push!(sequences, string(FASTA.sequence(record)))\n",
    "    end\n",
    "    close(reader)\n",
    "    \n",
    "    return identifiers, sequences\n",
    "end\n",
    "#read_input(\"input1.txt\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 9. Main function [Execute this function with your custom input in the \"input1.txt\" file in fasta format]\n",
    "\n",
    "The driver function implemented below reads input sequences from the \"input1.txt\" file, executes the NSGA algorithm and stores the optimal alignments back to the \"output1.txt\" file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input sequences:\n",
      "AATGGTCATAGCGAGATGAAGCCACGTGATGGATAATATTGTGCAAACGACCTTATTAGCTATTGACCGTCGATGTCCAACGAGACAATT\n",
      "GAATCTGTATTCTTCAAGCTTCAACTCCATGCACTACGAACGGTAGTGGTTCACATTGACCGTG\n",
      "TTGGGCGCATTGACCGTCCTTCCTAGCGTATCATCAAACTTGTGATTCTCTATCTAGAGCAAAATGCGGTGTCCGCTATATGGAGATCTATTTCAAAATA\n",
      "GTTACATATCAGAATCATTAGAAACGCTCTTAATGGGGTTAAGCAGAGACTTAGTAAGGATTAACTCCCAAGATGATTGACCGTGCTC\n",
      "\n",
      "Alignment process started:\n",
      "\n",
      "Generated alignments:\n",
      "AATGG-TCATAGCGAG-ATGAAGCCACGTGATGGATAATA----TTGTGCA-----AACGACC-TTATTAGCTATTGACCG----TCGATGTCCAACGAGACAATT\n",
      "GAATCTGTATTCTTC-AAGCTT-CAAC----T---CCATGC-------------------AC----T-------ACGAACGG---TAGTGGTTCACATTGACCGTG\n",
      "TTGGGCGCATTGACCG-TCCTTCCTAGCGTATC----ATCAAACTTGTGATTCTCTATCTAGAGC-AAAATGCGGTGTCCGCTATATGGAGATCTATTTCAAAATA\n",
      "GTTACAT-ATCA---GAATCATTAGAAACGCTCT-TAATG----GGGTTAAGC---A--GAGACTTAGTAAG-GATTAACTCCC-AAGATGATTGAC--CGTGCTC\n",
      "\n",
      "Sum of pairs scor[+1 for match, 0 otherwise]: 180.0\n",
      "\n",
      "Total time taken to align the sequence (in seconds): 64.0158182\n",
      "\n",
      "This alignment is stored in the output1.txt file\n"
     ]
    }
   ],
   "source": [
    "function generate_alignments(file_name)\n",
    "    identifiers, sequences = read_input(file_name)\n",
    "    population_size = 200\n",
    "    num_of_generations = 500\n",
    "    mutation_prob = 0.8\n",
    "    \n",
    "    println(\"Input sequences:\")\n",
    "    for each in sequences\n",
    "        println(each)\n",
    "    end\n",
    "    println()\n",
    "    \n",
    "    println(\"Alignment process started:\")\n",
    "    println()\n",
    "    \n",
    "    optimal_individual = @timed NSGA(sequences, population_size, num_of_generations, mutation_prob)\n",
    "    optimal_align = optimal_individual.value.align\n",
    "    score = optimal_individual.value.y[2]\n",
    "    \n",
    "    \n",
    "    println(\"Generated alignments:\")\n",
    "    for each in optimal_align\n",
    "        println(each)\n",
    "    end\n",
    "    println()\n",
    "    println(\"Sum of pairs scor[+1 for match, 0 otherwise]: \", score) \n",
    "    println()\n",
    "    println(\"Total time taken to align the sequence (in seconds): \", optimal_individual.time)\n",
    "    println()\n",
    "    \n",
    "    open(\"output1.txt\", \"w\") do file\n",
    "        for i in 1:length(optimal_align)\n",
    "            write(file, \">\"*identifiers[i]*\"\\n\")\n",
    "            write(file, optimal_align[i]*\"\\n\")\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    println(\"This alignment is stored in the output1.txt file\")\n",
    "    \n",
    "end\n",
    "generate_alignments(\"input1.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Algorithm performance experiments\n",
    "\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "generate_dna_sequences (generic function with 1 method)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function generate_dna_sequences(t,l)\n",
    "    \"\"\"\n",
    "    This function generated the random DNA Sequences for the experiments conducted below.\n",
    "    \"\"\"\n",
    "    dna_sequences=[]\n",
    "    for i in 1:t\n",
    "        dna=randstring(\"AGTC\", l)\n",
    "        push!(dna_sequences,dna)\n",
    "    end\n",
    "    return dna_sequences\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NSGA_Experiment (generic function with 1 method)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function NSGA_Experiment(input_seq, population_size, num_of_gen, pmut)\n",
    "    \"\"\"\n",
    "    This is a little modification to above implemented NSGA function which reports the results after certain number of generations for the experiments purpose.\n",
    "    \"\"\"\n",
    "    alignments, codes = generate_initial_population(input_seq, population_size)\n",
    "    P = Vector{Individual}(undef, 2*population_size)    # Whole population \n",
    "    for i in 1:population_size                                # Create Initial population\n",
    "        P[i] = Individual(alignments[i],codes[i], calculate_fitness_score(alignments[i])) \n",
    "        P[population_size+i] = deepcopy(P[i])\n",
    "    end\n",
    "    calculate_Pareto_Fronts!(view(P, 1:population_size))\n",
    "    generations=[]\n",
    "    align_score=[]\n",
    "    for g in 50:20:250\n",
    "        append!(generations,g)\n",
    "    end\n",
    "    for gen in 1:num_of_gen\n",
    "        for i = 1:2:population_size\n",
    "            pa = tournament_selection(P)\n",
    "            pb = tournament_selection(P)\n",
    "            childs_align, childs_codification = crossover(pa.align, pa.codification, pb.align, pb.codification)\n",
    "            if rand() < pmut\n",
    "                childs_align[1],childs_codification[1] = mutation(childs_align[1],childs_codification[1])\n",
    "            end\n",
    "            if rand() < pmut\n",
    "                childs_align[2], childs_codification[2] = mutation(childs_align[2],childs_codification[2])\n",
    "            end\n",
    "            P[population_size+i] = Individual(childs_align[1],childs_codification[1], calculate_fitness_score(childs_align[1])) \n",
    "            P[population_size+i+1] = Individual(childs_align[2],childs_codification[2], calculate_fitness_score(childs_align[2])) \n",
    "        end\n",
    "        calculate_Pareto_Fronts!(P)\n",
    "        sort!(P, by = x -> x.rank, alg = Base.Sort.QuickSort)\n",
    "        let f::Int = 1\n",
    "            ind = 0\n",
    "            indnext = findlast(x -> x.rank == f, P)\n",
    "            while 0 < indnext <= population_size\n",
    "                ind = indnext\n",
    "                f += 1\n",
    "                indnext = findlast(x -> x.rank == f, P)\n",
    "            end\n",
    "            indnext == 0 && (indnext = length(P))\n",
    "            crowding_distance_sorting!(view(P, ind+1:indnext))\n",
    "            sort!(view(P, (ind + 1):indnext), by = x -> x.crowding, rev = true, alg = PartialQuickSort(population_size - ind))\n",
    "        end\n",
    "        if gen in generations\n",
    "            optimal_sols = filter(x -> x.rank == 1, view(P, 1:population_size))\n",
    "            best_sol = optimal_sols[1]\n",
    "            for i in 2:length(optimal_sols)\n",
    "                if optimal_sols[i].y[2] > best_sol.y[2]\n",
    "                     best_sol = optimal_sols[i]\n",
    "                end \n",
    "            end\n",
    "            append!(align_score,best_sol.y[2])\n",
    "        end\n",
    "    end\n",
    "    return generations, align_score\n",
    "end\n",
    "#input_seq = [\"AATGGTCATAGCGAGATGAAGCCACGTGATGGATAATATTGTGCAAACGACCTTATTAGCTATTGACCGTCGATGTCCAACGAGACAATT\", \"GAATCTGTATTCTTCAAGCTTCAACTCCATGCACTACGAACGGTAGTGGTTCACATTGACCGTG\", \"TTGGGCGCATTGACCGTCCTTCCTAGCGTATCATCAAACTTGTGATTCTCTATCTAGAGCAAAATGCGGTGTCCGCTATATGGAGATCTATTTCAAAATA\", \"GTTACATATCAGAATCATTAGAAACGCTCTTAATGGGGTTAAGCAGAGACTTAGTAAGGATTAACTCCCAAGATGATTGACCGTGCTC\"]\n",
    "#NSGA(input_seq,100,300,0.8)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot Graphs\n",
    "\n",
    "**Plot 1:** Time taken by NSGA 2 algorithm with respect to number of generations.\n",
    "\n",
    "**Plot 2:** Resulting alignment score by NSGA 2 algorithm with respect to number of generations.\n",
    "\n",
    "**Plot 3:** Time taken by NSGA 2 algorithm with respect to number of sequences.\n",
    "\n",
    "**Plot 4:** Memory consumed by NSGA 2 algorithm with respect to number of sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "plot_graphs (generic function with 1 method)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function plot_graphs()\n",
    "    \"\"\"\n",
    "    This function plots the above mentioned four graphs in order to assess the performance of the algorithm.\n",
    "    \"\"\"\n",
    "    population_size = 100\n",
    "    num_of_generations = 250\n",
    "    mutation_prob = 0.2\n",
    "    dna_length = 20\n",
    "    num_of_sequences = 10\n",
    "    time_taken=[]\n",
    "    sequences = generate_dna_sequences(num_of_sequences,dna_length)\n",
    "    generations, align_score = NSGA_Experiment(sequences, population_size, num_of_generations,mutation_prob)\n",
    "    for g in 50:20:250\n",
    "        curr_time=0\n",
    "        curr_time = @timed begin\n",
    "            best_sol = NSGA(sequences, population_size, g, mutation_prob)\n",
    "        end\n",
    "        append!(time_taken,curr_time[2])\n",
    "    end\n",
    "    sequence_time=[]\n",
    "    sequence_memory=[]\n",
    "    sequence_length=[]\n",
    "    for s in 10:2:30\n",
    "        sequences = generate_dna_sequences(num_of_sequences,s)\n",
    "        time_memory = @timed begin\n",
    "            NSGA(sequences, population_size, num_of_generations, mutation_prob)\n",
    "        end\n",
    "        append!(sequence_time, time_memory[2])\n",
    "        append!(sequence_memory, (time_memory[3]/1000000))\n",
    "        append!(sequence_length, s)\n",
    "    end\n",
    "    p1 = plot( generations, time_taken, ylabel = \"Time taken (in seconds)\", xlabel = \"No. of Generations\", title = \"Generations Vs Time\")\n",
    "    p2 = plot(generations, align_score, ylabel = \"Alignment Score (SOP score)\", xlabel = \"No. of Generations\", title = \"Generations vs Alignment Score\")\n",
    "    p3 = plot(sequence_length,sequence_time, ylabel = \"Time taken (in seconds)\", xlabel = \"Sequence length\", title = \"Sequence length Vs Time\")\n",
    "    p4 = plot(sequence_length, sequence_memory, ylabel = \"Memory (in MB)\", xlabel = \"Sequence length\", title = \"Sequence length Vs Memory\")\n",
    "    display(p1)\n",
    "    display(p2)\n",
    "    display(p3)\n",
    "    display(p4)\n",
    "end\n",
    "#plot_graphs()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running algorithm on the BAliBASE dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NSGA_for_balibase (generic function with 1 method)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function generate_initial_population_from_align(population_count,alignments) \n",
    "    without_crossover = length(alignments)\n",
    "    \n",
    "    population_string = []\n",
    "    population_coded = []\n",
    "    \n",
    "    \n",
    "    for i in 1:without_crossover\n",
    "        coded_chromosome = codify_chromosome(alignments[i])\n",
    "        push!(population_coded,coded_chromosome)\n",
    "        push!(population_string,alignments[i])\n",
    "    end\n",
    "    \n",
    "    with_crossover = (population_count - without_crossover)/2\n",
    "    for i in 1:with_crossover\n",
    "        random1 = Int(rand(1:without_crossover))\n",
    "        random2 = Int(rand(1:without_crossover))\n",
    "        crossover_string, crossover_coded = crossover(population_string[random1],population_coded[random1],population_string[random2],population_coded[random2])\n",
    "        population_string = vcat(population_string,crossover_string)\n",
    "        population_coded = vcat(population_coded,crossover_coded)\n",
    "    end\n",
    "    \n",
    "    return population_string, population_coded\n",
    "end   \n",
    "\n",
    "\n",
    "function NSGA_for_balibase(input_alignments, population_size, num_of_gen, pmut)\n",
    "    alignments, codes = generate_initial_population_from_align(population_size,input_alignments)\n",
    "    \n",
    "    P = Vector{Individual}(undef, 2*population_size)    # Whole population \n",
    "    \n",
    "    for i in 1:population_size                                # Create Initial population\n",
    "        P[i] = Individual(alignments[i],codes[i], calculate_fitness_score(alignments[i])) \n",
    "        P[population_size+i] = deepcopy(P[i])\n",
    "    end\n",
    "    \n",
    "    calculate_Pareto_Fronts!(view(P, 1:population_size))\n",
    "    \n",
    "    for gen in 1:num_of_gen\n",
    "        for i = 1:2:population_size\n",
    "            \n",
    "            pa = tournament_selection(P)\n",
    "            pb = tournament_selection(P)\n",
    "            \n",
    "            childs_align, childs_codification = crossover(pa.align, pa.codification, pb.align, pb.codification)\n",
    "            \n",
    "            if rand() < pmut\n",
    "                childs_align[1],childs_codification[1] = mutation(childs_align[1],childs_codification[1])\n",
    "            end\n",
    "                \n",
    "            if rand() < pmut\n",
    "                childs_align[2], childs_codification[2] = mutation(childs_align[2],childs_codification[2])\n",
    "            end\n",
    "            \n",
    "            P[population_size+i] = Individual(childs_align[1],childs_codification[1], calculate_fitness_score(childs_align[1])) \n",
    "            P[population_size+i+1] = Individual(childs_align[2],childs_codification[2], calculate_fitness_score(childs_align[2])) \n",
    "            \n",
    "        end\n",
    "        calculate_Pareto_Fronts!(P)\n",
    "        \n",
    "        sort!(P, by = x -> x.rank, alg = Base.Sort.QuickSort)\n",
    "        \n",
    "        let f::Int = 1\n",
    "            ind = 0\n",
    "            indnext = findlast(x -> x.rank == f, P)\n",
    "            while 0 < indnext <= population_size\n",
    "                ind = indnext\n",
    "                f += 1\n",
    "                indnext = findlast(x -> x.rank == f, P)\n",
    "            end\n",
    "            indnext == 0 && (indnext = length(P))\n",
    "            crowding_distance_sorting!(view(P, ind+1:indnext))\n",
    "            sort!(view(P, (ind + 1):indnext), by = x -> x.crowding, rev = true, alg = PartialQuickSort(population_size - ind))\n",
    "            \n",
    "        end\n",
    "        \n",
    "    end\n",
    "    optimal_sols = filter(x -> x.rank == 1, view(P, 1:population_size))\n",
    "    \n",
    "    best_sol = optimal_sols[1]\n",
    "    for i in 2:length(optimal_sols)\n",
    "        if optimal_sols[i].y[2] > best_sol.y[2]\n",
    "             best_sol = optimal_sols[i]\n",
    "        end \n",
    "    end\n",
    "    \n",
    "    return best_sol\n",
    "    \n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "experiment (generic function with 1 method)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Experiments for the result section\n",
    "\n",
    "function experiment(version_name)\n",
    "    release_dir = \"bb3_dataset\\\\\"*version_name\n",
    "    align_dir = \"bb3_aligned\\\\\"*version_name\n",
    "    \n",
    "    files = readdir(release_dir)\n",
    "    \n",
    "    aligned_files = [\".msf_tfa\",\".tfa_clu\",\".tfa_fsa\",\".tfa_kalign\",\".tfa_mafft\",\".tfa_muscle\",\".tfa_probcons\",\".tfa_retalign\"]\n",
    "\n",
    "    \n",
    "    population_size = 100\n",
    "    num_of_generations = 300\n",
    "    mutation_prob = 0.5\n",
    "    \n",
    "    for file_name in files\n",
    "        if file_name[end-3:end] != \".tfa\"\n",
    "           continue \n",
    "        end\n",
    "        \n",
    "        alignments = []\n",
    "        \n",
    "        for af in aligned_files\n",
    "            name = align_dir*\"\\\\\"*file_name[1:end-4]*af\n",
    "            _, sequences = read_input(name)\n",
    "            push!(alignments, sequences)\n",
    "            score = calc_sum_pair(sequences)\n",
    "            println(file_name[1:end-4]*af, \" : \", score)\n",
    "        end\n",
    "        \n",
    "        optimal_align = NSGA_for_balibase(alignments, population_size, num_of_generations, mutation_prob).align\n",
    "        score = calc_sum_pair(optimal_align)\n",
    "        \n",
    "        println(file_name, \" : \", round(score, digits=4))\n",
    "        \n",
    "        println(\"-----------------------------------------------------\")\n",
    "        \n",
    "    end\n",
    "    \n",
    "end\n",
    "\n",
    "#experiment(\"RV11\")\n",
    "#experiment(\"RV12\")\n",
    "#experiment(\"RV20\")\n",
    "#experiment(\"RV30\")\n",
    "#experiment(\"RV40\")\n",
    "#experiment(\"RV50\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.5.0",
   "language": "julia",
   "name": "julia-1.5"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.5.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
